crates/typst/src/text/font/color.rs

```diff
@@ -0,0 +1,321 @@
+    frame.push(position, FrameItem::Image(image, size, Span::detached()));
+}
+
+/// Draws a COLR glyph in a frame.
+fn draw_colr_glyph(frame: &mut Frame, font: &Font, fill: &Paint, glyph_id: GlyphId) {
+    let mut painter = ColrPainter { font, fill, current_glyph: glyph_id, frame };
+    font.ttf().paint_color_glyph(glyph_id, 0, &mut painter);
+}
+
+/// Draws an SVG glyph in a frame.
+fn draw_svg_glyph(frame: &mut Frame, upem: Abs, font: &Font, glyph_id: GlyphId) {
+    let Some(SizedSvg { tree, bbox }) = get_svg_glyph(font, glyph_id) else {
+        // Don't draw anything in case we were not able to parse and measure the
+        // SVG.
+        return;
+    };
+
+    let mut data = tree.to_string(&usvg::XmlOptions::default());
+
+    let width = bbox.width() as f64;
+    let height = bbox.height() as f64;
+    let left = bbox.left() as f64;
+    let top = bbox.top() as f64;
+
+    // The SVG coordinates and the font coordinates are not the same: the Y axis
+    // is mirrored. But the origin of the axes are the same (which means that
+    // the horizontal axis in the SVG document corresponds to the baseline). See
+    // the reference for more details:
+    // https://learn.microsoft.com/en-us/typography/opentype/spec/svg#coordinate-systems-and-glyph-metrics
+    //
+    // If we used the SVG document as it is, svg2pdf would produce a cropped
+    // glyph (only what is under the baseline would be visible). So we need to
+    // embed the original SVG in another one that has the exact dimensions of
+    // the glyph, with a transform to make it fit. We also need to remove the
+    // viewBox, height and width attributes from the inner SVG, otherwise usvg
+    // takes into account these values to clip the embedded SVG.
+    make_svg_unsized(&mut data);
+    let wrapper_svg = format!(
+        r#"
+        <svg
+            width="{width}"
+            height="{height}"
+            viewBox="0 0 {width} {height}"
+            xmlns="http://www.w3.org/2000/svg">
+            <g transform="matrix(1 0 0 1 {tx} {ty})">
+            {inner}
+            </g>
+        </svg>
+    "#,
+        inner = data,
+        tx = -left,
+        ty = -top,
+    );
+
+    let image = Image::new(
+        wrapper_svg.as_bytes().into(),
+        typst::visualize::ImageFormat::Vector(typst::visualize::VectorFormat::Svg),
+        None,
+    )
+    .unwrap();
+    let position = Point::new(Abs::pt(left), Abs::pt(top) + upem);
+    let size = Axes::new(Abs::pt(width), Abs::pt(height));
+    frame.push(position, FrameItem::Image(image, size, Span::detached()));
+}
+
+/// Remove all size specifications (viewBox, width and height attributes) from a
+/// SVG document.
+fn make_svg_unsized(svg: &mut String) {
+    let mut viewbox_range = None;
+    let mut width_range = None;
+    let mut height_range = None;
+
+    let mut s = unscanny::Scanner::new(svg);
+
+    s.eat_until("<svg");
+    s.eat_if("<svg");
+    while !s.eat_if('>') {
```

Comment:
this could potentially put Typst into an infinite loop, this should also check `s.done()` or do something else. same below.